File: cine/AMarotoChivite-Cine/AMarotoChivite-Cine-Java/src/main/java/simulacionCine/utils/funcionesMenuAdmin.java
package simulacionCine.utils;

import simulacionCine.enums.EstadoButaca;
import simulacionCine.models.Butaca;
import simulacionCine.models.Cliente;
import simulacionCine.models.Sala;

import java.util.Scanner;

import static simulacionCine.simulacionCine.*;
import static simulacionCine.utils.funcionesEsteticaDeConsola.*;


// import static kotlin.io.ConsoleKt.readln;

public class funcionesMenuAdmin {

    /**
     * Verifica si una contraseña es correcta.
     *
     * @param passCorrect La contraseña correcta a comparar.
     * @return "true" si la contraseña es correcta, "false" si no lo es.
     */
    public static boolean isPassCorrect(String passCorrect) {
        while (true) {
            System.out.println("Introduzca la contraseña:");
            // Importando librería de kotlin
            //String entradaInicioSesion = readln();

            // Con puro JAVA
            Scanner sc = new Scanner(System.in);
            String entradaInicioSesion = sc.nextLine();

            if (entradaInicioSesion.equals(passCorrect)) {
                falsoBorradoDeConsola();
                return true;
            } else {
                System.out.println("Contraseña incorrecta!");
                try {
                    Thread.sleep(1500);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                falsoBorradoDeConsola();
                return false;
            }
        }
    }

    /**
     * Muestra un menú para el administrador.
     *
     * @param cine El cine del que se desea obtener información.
     */
    public static void menuAdmin(Sala[] cine, Cliente[] almacenClientes) {
        while (true) {
            portadaCinesAngel();
            System.out.println("(ADMIN) -> Selecciona la opción deseada: ");
            System.out.println();
            System.out.println("1: Informe de butacas (libres, reservadas y ocupadas)");
            System.out.println("2: Recaudación total (caja final de las butacas compradas/ocupadas)");
            System.out.println("0: Salir");

            Scanner sc = new Scanner(System.in);
            switch (sc.nextLine()) {
                // Admin
                case "1" -> {
                    falsoBorradoDeConsola();
                    informeButacas(cine);
                    while (!volverAlMenu()) {
                        // No saldremos del bucle hasta que introduzcamos "menu"
                    }
                    falsoBorradoDeConsola();
                }
                case "2" -> {
                    falsoBorradoDeConsola();
                    informeRecaudacionTotal(cine);
                    while (!volverAlMenu()) {
                        // No saldremos del bucle hasta que introduzcamos "menu"
                    }
                    falsoBorradoDeConsola();

                }
                // Salir
                case "0" -> {
                    falsoBorradoDeConsola();
                    menuEleccionUsuario(cine, almacenClientes);
                }
            }
        }
    }

    /**
     * Muestra un informe de las butacas de una sala determinada.
     *
     * @param cine El cine del que se desea obtener información.
     */
    public static void informeButacas(Sala[] cine) {
        String[] almacenIDSalaElegida = new String[1];
        almacenIDSalaElegida[0] = " ";

        boolean salirBucle = false;
        while (!salirBucle) {
            mostrarCatalogo(cine);
            System.out.println("Introduzca el ID de la sala para representar un informe de su estado:");
            Scanner sc = new Scanner(System.in);
            String entradaIDsala = sc.nextLine();
            falsoBorradoDeConsola();

            // Filtro para elegir un ID existente
            for (int i = 0; i < cine.length; i++) {
                if (entradaIDsala.equals(cine[i].id)) {
                    almacenIDSalaElegida[0] = entradaIDsala;
                    salirBucle = true;
                    break;
                }
                if (i == cine.length - 1) {
                    if (!entradaIDsala.equals(cine[i].id)) {
                        System.out.println("ID inválido! Debe existir la sala en el CINE!");
                        try {
                            Thread.sleep(2250);
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                        falsoBorradoDeConsola();
                        break;
                    }
                }

            }
        }

        mostrarSala(cine, almacenIDSalaElegida[0]);


        Sala[] salaElegida = new Sala[1];
        salaElegida[0] = cine[0];
        for (int i = 0; i < cine.length; i++) {
            if (cine[i].id.equals(almacenIDSalaElegida[0])) {
                salaElegida[0] = cine[i];
            }
        }

        Butaca[][] sala = salaElegida[0].getMatrizSala();

        int contadorButacasLibres = 0;
        int contadorButacasLibresVIP = 0;
        int contadorButacasReservadas = 0;
        int contadorButacasReservadasVIP = 0;
        int contadorButacasOcupadas = 0;
        int contadorButacasOcupadasVIP = 0;

        for (int filas = 0; filas < sala.length; filas++) {
            for (int columnas = 0; columnas < sala[filas].length; columnas++) {
                // Si es ESTÁNDAR libres
                if (sala[filas][columnas].getEstadoButaca() == EstadoButaca.LIBRE && !sala[filas][columnas].getBooleanButacaVip()) {
                    contadorButacasLibres += 1;
                }
                // Si es VIP libres
                if (sala[filas][columnas].getEstadoButaca() == EstadoButaca.LIBRE && sala[filas][columnas].getBooleanButacaVip()) {
                    contadorButacasLibresVIP += 1;
                }
                // Si es ESTÁNDAR reservadas
                if (sala[filas][columnas].getEstadoButaca() == EstadoButaca.RESERVADO && !sala[filas][columnas].getBooleanButacaVip()) {
                    contadorButacasReservadas += 1;
                }
                // Si es VIP reservadas
                if (sala[filas][columnas].getEstadoButaca() == EstadoButaca.RESERVADO && sala[filas][columnas].getBooleanButacaVip()) {
                    contadorButacasReservadasVIP += 1;
                }
                // Si es ESTÁNDAR ocupadas
                if (sala[filas][columnas].getEstadoButaca() == EstadoButaca.OCUPADO && !sala[filas][columnas].getBooleanButacaVip()) {
                    contadorButacasOcupadas += 1;
                }
                // Si es VIP ocupadas
                if (sala[filas][columnas].getEstadoButaca() == EstadoButaca.OCUPADO && sala[filas][columnas].getBooleanButacaVip()) {
                    contadorButacasOcupadasVIP += 1;
                }
            }
        }
        System.out.println("====================Informe SALA " + salaElegida[0].id + "=======================");
        System.out.println("Butacas ESTÁNDAR libres -> (" + contadorButacasLibres + ")");
        System.out.println("Butacas VIP libres ->  (" + contadorButacasLibresVIP + ")");
        System.out.println();
        System.out.println("Butacas ESTÁNDAR reservadas -> (" + contadorButacasReservadas + ")");
        System.out.println("Butacas VIP reservadas ->  (" + contadorButacasReservadasVIP + ")");
        System.out.println();
        System.out.println("Butacas ESTÁNDAR ocupadas ->(" + contadorButacasOcupadas + ")");
        System.out.println("Butacas VIP ocupadas ->  (" + contadorButacasOcupadasVIP + ")");
        System.out.println("=========================================================");
    }

    /**
     * Calcula la recaudación total del cine en función número de butacas ocupadas y su precio.
     *
     * @param cine Array de objetos Sala con las salas del cine
     */
    static void informeRecaudacionTotal(Sala[] cine) {
        int contadorButacasVIP = 0;
        int contadorButacasNoVIP = 0;

        // Tenemos que recorrer cada sala del cine
        for (int i = 0; i < cine.length; i++) {
            Butaca[][] sala = cine[i].getMatrizSala();
            // Recorremos cada sala
            for (int filas = 0; filas < sala.length; filas++) {
                for (int columnas = 0; columnas < sala[filas].length; columnas++) {
                    // Si es VIP
                    if (sala[filas][columnas].getEstadoButaca() == EstadoButaca.OCUPADO && sala[filas][columnas].getBooleanButacaVip()) {
                        contadorButacasVIP++;
                    }
                    // Si no es VIP
                    if (sala[filas][columnas].getEstadoButaca() == EstadoButaca.OCUPADO && !sala[filas][columnas].getBooleanButacaVip()) {
                        contadorButacasNoVIP++;
                    }
                }
            }
        }

        double totalVIP = contadorButacasVIP * PRECIO_VIP;
        double totalNoVIP = contadorButacasNoVIP * PRECIO_ESTANDAR;
        double totalFinal = totalVIP + totalNoVIP;
        System.out.println("====================Informe de CAJA FINAL=======================");
        System.out.println("Recaudación butacas ESTÁNDAR (" + contadorButacasNoVIP + ") -> " + totalNoVIP + "€ ");
        System.out.println("Recaudación butacas VIP (" + contadorButacasVIP + ") -> " + totalVIP + "€ ");
        System.out.println("Recaudación TOTAL -> " + totalFinal + "€ ");
        System.out.println("================================================================");
    }
}

File: cine/AMarotoChivite-Cine/AMarotoChivite-Cine-Kotlin/src/main/kotlin/simulacionCine/utils/funcionesMenuAdmin.kt
package simulacionCine.utils

import simulacionCine.PRECIO_ESTANDAR
import simulacionCine.PRECIO_VIP
import simulacionCine.enums.Color
import simulacionCine.enums.EstadoButaca
import simulacionCine.models.Butaca
import simulacionCine.models.Sala
import simulacionCine.volverAlMenu

/**
 * Muestra un menú para el administrador.
 *
 * @param cine El cine del que se desea obtener información.
 */
fun menuAdmin(cine: Array<Sala>) {
    while (true) {
        portadaCinesAngel()
        println("(${Color.PURPLE_BRIGHT.color}ADMIN${Color.RESET.color}) -> Selecciona la opción deseada: ")
        println("")
        println("1: Informe de butacas (libres, reservadas y ocupadas)")
        println("2: Recaudación total (caja final de las butacas compradas/ocupadas)")
        println("0: Salir")

        when (readln()) {
            // Admin
            "1" -> {
                falsoBorradoDeConsola()
                informeButacas(cine)
                while (!volverAlMenu()) {
                    // No saldremos del bucle hasta que introduzcamos "menu"
                }
                falsoBorradoDeConsola()
            }

            "2" -> {
                falsoBorradoDeConsola()
                informeRecaudacionTotal(cine)
                while (!volverAlMenu()) {
                    // No saldremos del bucle hasta que introduzcamos "menu"
                }
                falsoBorradoDeConsola()
            }
            // Salir
            "0" -> {
                falsoBorradoDeConsola()
                break
            }
        }
    }
}

/**
 * Muestra un informe de las butacas de una sala determinada.
 *
 * @param cine El cine del que se desea obtener información.
 */
fun informeButacas(cine: Array<Sala>) {
    val almacenIDSalaElegida: Array<String> = Array<String>(1) { " " }

    var salirBucle: Boolean = false
    while (!salirBucle) {
        mostrarCatalogo(cine)
        println("Introduzca el ID de la sala para representar un informe de su estado:")
        val entradaIDsala: String = readln()
        falsoBorradoDeConsola()

        // Filtro para elegir un ID existente
        for (i in cine.indices) {
            if (entradaIDsala == cine[i].id) {
                almacenIDSalaElegida[0] = entradaIDsala
                salirBucle = true
                break
            }
            if (i == cine.size - 1) {
                if (entradaIDsala != cine[i].id) {
                    println("ID inválido! Debe existir la sala en el CINE!")
                    Thread.sleep(2250)
                    falsoBorradoDeConsola()
                    break
                }
            }

        }

    }

    mostrarSala(cine, almacenIDSalaElegida[0])


    val salaElegida: Array<Sala> = Array<Sala>(1) { cine[0] }
    for (i in cine.indices) {
        if (cine[i].id == almacenIDSalaElegida[0]) {
            salaElegida[0] = cine[i]
        }
    }

    val sala: Array<Array<Butaca>> = salaElegida[0].getMatrizSala()

    var contadorButacasLibres: Int = 0
    var contadorButacasLibresVIP: Int = 0
    var contadorButacasReservadas: Int = 0
    var contadorButacasReservadasVIP: Int = 0
    var contadorButacasOcupadas: Int = 0
    var contadorButacasOcupadasVIP: Int = 0

    for (filas in sala.indices) {
        for (columnas in 0 until sala[filas].size) {
            // Si es ESTÁNDAR libres
            if (sala[filas][columnas].getEstadoButaca() == EstadoButaca.LIBRE && !sala[filas][columnas].getBooleanButacaVip()) {
                contadorButacasLibres += 1
            }
            // Si es VIP libres
            if (sala[filas][columnas].getEstadoButaca() == EstadoButaca.LIBRE && sala[filas][columnas].getBooleanButacaVip()) {
                contadorButacasLibresVIP += 1
            }
            // Si es ESTÁNDAR reservadas
            if (sala[filas][columnas].getEstadoButaca() == EstadoButaca.RESERVADO && !sala[filas][columnas].getBooleanButacaVip()) {
                contadorButacasReservadas += 1
            }
            // Si es VIP reservadas
            if (sala[filas][columnas].getEstadoButaca() == EstadoButaca.RESERVADO && sala[filas][columnas].getBooleanButacaVip()) {
                contadorButacasReservadasVIP += 1
            }
            // Si es ESTÁNDAR ocupadas
            if (sala[filas][columnas].getEstadoButaca() == EstadoButaca.OCUPADO && !sala[filas][columnas].getBooleanButacaVip()) {
                contadorButacasOcupadas += 1
            }
            // Si es VIP ocupadas
            if (sala[filas][columnas].getEstadoButaca() == EstadoButaca.OCUPADO && sala[filas][columnas].getBooleanButacaVip()) {
                contadorButacasOcupadasVIP += 1
            }

        }
    }
    println("====================${Color.CYAN_BRIGHT.color}Informe SALA ${salaElegida[0].id}${Color.RESET.color}=======================")
    println("Butacas ${Color.GREEN.color}ESTÁNDAR libres${Color.RESET.color} -> (${Color.GREEN.color}$contadorButacasLibres${Color.RESET.color})")
    println("Butacas ${Color.PURPLE_BRIGHT.color}VIP libres${Color.RESET.color} ->  (${Color.PURPLE_BRIGHT.color}$contadorButacasLibresVIP${Color.RESET.color})")
    println("")
    println("Butacas ${Color.YELLOW_BRIGHT.color}ESTÁNDAR reservadas${Color.RESET.color} -> (${Color.YELLOW_BRIGHT.color}$contadorButacasReservadas${Color.RESET.color})")
    println("Butacas ${Color.PURPLE_BRIGHT.color}VIP reservadas${Color.RESET.color} -> (${Color.PURPLE_BRIGHT.color}$contadorButacasReservadasVIP${Color.RESET.color})")
    println("")
    println("Butacas ${Color.RED_BRIGHT.color}ESTÁNDAR ocupadas${Color.RESET.color} -> (${Color.RED_BRIGHT.color}$contadorButacasOcupadas${Color.RESET.color})")
    println("Butacas ${Color.PURPLE_BRIGHT.color}VIP ocupadas${Color.RESET.color} ->  (${Color.PURPLE_BRIGHT.color}$contadorButacasOcupadasVIP${Color.RESET.color})")
    println("=========================================================")
}

/**
 * Calcula la recaudación total del cine en función número de butacas ocupadas y su precio.
 *
 * @param cine Array de objetos Sala con las salas del cine
 */
fun informeRecaudacionTotal(cine: Array<Sala>) {
    var contadorButacasVIP: Int = 0
    var contadorButacasNoVIP: Int = 0

    // Tenemos que recorrer cada sala del cine
    for (i in cine.indices) {
        var sala: Array<Array<Butaca>> = cine[i].getMatrizSala()
        // Recorremos cada sala
        for (filas in sala.indices) {
            for (columnas in 0 until sala[filas].size) {
                // Si es VIP
                if (sala[filas][columnas].getEstadoButaca() == EstadoButaca.OCUPADO && sala[filas][columnas].getBooleanButacaVip()) {
                    contadorButacasVIP += 1
                }
                // Si no es VIP
                if (sala[filas][columnas].getEstadoButaca() == EstadoButaca.OCUPADO && !sala[filas][columnas].getBooleanButacaVip()) {
                    contadorButacasNoVIP += 1
                }
            }
        }
    }

    val totalVIP: Double = contadorButacasVIP * PRECIO_VIP
    val totalNoVIP: Double = contadorButacasNoVIP * PRECIO_ESTANDAR
    val totalFinal: Double = totalVIP + totalNoVIP
    println("====================${Color.CYAN_BRIGHT.color}Informe de CAJA FINAL${Color.RESET.color}=======================")
    println("Recaudación butacas ${Color.BLUE.color}ESTÁNDAR${Color.RESET.color} (${Color.BLUE.color}$contadorButacasNoVIP${Color.RESET.color}) -> ${Color.BLUE.color}$totalNoVIP€${Color.RESET.color} ")
    println("Recaudación butacas ${Color.PURPLE_BRIGHT.color}VIP${Color.RESET.color} (${Color.PURPLE_BRIGHT.color}$contadorButacasVIP${Color.RESET.color}) -> ${Color.PURPLE_BRIGHT.color}$totalVIP€${Color.RESET.color} ")
    println("Recaudación ${Color.CYAN_BRIGHT.color}TOTAL${Color.RESET.color} -> ${Color.CYAN_BRIGHT.color}$totalFinal€${Color.RESET.color} ")
    println("================================================================")
}

/**
 * Verifica si una contraseña es correcta.
 *
 * @param passCorrect La contraseña correcta a comparar.
 * @return "true" si la contraseña es correcta, "false" si no lo es.
 */
fun isPassCorrect(passCorrect: String): Boolean {
    while (true) {
        println("Introduzca la contraseña:")
        val entradaInicioSesion: String = readln()
        if (entradaInicioSesion == passCorrect) {
            falsoBorradoDeConsola()
            return true
        } else {
            println("Contraseña incorrecta!")
            Thread.sleep(1500)
            falsoBorradoDeConsola()
            return false
        }
    }
}